### DISCUSSION TITLE
Scipy error when creating template

### DISCUSSION THREAD

**Discussion by kovtunlab** [SBC-Utrecht/pytom-match-pick#172] [Questions]
Hello, 

I encountered a potential bug. 

The TM works as described for the tutorial inputs, but crashes on attempt to process my reference witth pytom_create_template.py
The command and error are   
$pytom_create_template.py -i arch_bin4_sharp.mrc -o arch_bin8.mrc --input-voxel-size 6.28 --output-voxel-size 12.56 --center -b 40 
pytom_tm/lib/python3.12/site-packages/scipy/ndimage/_measurements.py:1541: RuntimeWarning: invalid value encountered in scalar divide
  results = [sum(input * grids[dir].astype(float), labels, index) / normalizer
INFO:root:Convoluting volume with filter and then downsampling.

I tried different ways of the mrc input file treatment (resaving it with e2proc3d.py eman2 or chimera), inverted or not-inverted, sharpened and non sharpened - same results.  
The values range in the mrc do not look different form the tutorial map:

$ header arch_bin4_sharp.mrc 

 RO image file on unit   1 : arch_bin4_sharp.mrc     Size=       1459 K

 Number of columns, rows, sections .....      72      72      72
 Map mode ..............................    2   (32-bit float)             
 Start cols, rows, sects, grid x,y,z ...    0     0     0      72     72     72
 Pixel spacing (Angstroms)..............   6.280      6.280      6.280    
 Cell angles ...........................   90.000   90.000   90.000
 Fast, medium, slow axes ...............    X    Y    Z
 Origin on x,y,z .......................    0.000       0.000       0.000    
 Minimum density ....................... -0.75089    
 Maximum density .......................   1.2693    
 Mean density ..........................  0.22659E-02
 tilt angles (original,current) ........   0.0   0.0   0.0   0.0   0.0   0.0
 Space group,# extra bytes,idtype,lens .        0        0        0        0

     1 Titles :
The tutorial map for comparison: 

header guidetest/emd_2938.mrc

 RO image file on unit   1 : guidetest/emd_2938.mrc     Size=     105469 K

 Number of columns, rows, sections .....     300     300     300
 Map mode ..............................    2   (32-bit float)             
 Start cols, rows, sects, grid x,y,z ...    0     0     0     300    300    300
 Pixel spacing (Angstroms)..............   1.100      1.100      1.100    
 Cell angles ...........................   90.000   90.000   90.000
 Fast, medium, slow axes ...............    X    Y    Z
 Origin on x,y,z .......................    0.000       0.000       0.000    
 Minimum density ....................... -0.23737    
 Maximum density .......................  0.36217    
 Mean density ..........................  0.75577E-03
 tilt angles (original,current) ........   0.0   0.0   0.0   0.0   0.0   0.0
 Space group,# extra bytes,idtype,lens .        0        0        0        0

     1 Titles :
Chimera 1.17_b42434 Mon Apr 29 17:19:00 2024                                   

Best. 
Oleksiy

**Comment by McHaillet** [comment]
Hi Oleksiy,

Thanks for reaching out! 

> pytom_tm/lib/python3.12/site-packages/scipy/ndimage/_measurements.py:1541: RuntimeWarning: invalid value encountered in scalar divide
results = [sum(input * grids[dir].astype(float), labels, index) / normalizer

It seems scipy is only throwing a warning, do you get the expected output?

Could try running the command without the `--center` option? Just for me to figure out if the warning originates from centering or rescaling, which is what scipy.ndimage is used for.

Best,
Marten

**Reply by sroet** [reply]
~Just to add to that: I believe the correct input output options are `--input-voxel-size-angstrom` and `--output-voxel-size-angstrom` for `create_template`  ever since the `0.2.0` release (didn't check further back). Which version of the code are you running?~

Nevermind, learned that argparse accepts unique shortened names as well

**Reply by kovtunlab** [reply]
cheers

**Comment by kovtunlab** [comment]
Hi Marten, 
Thank you for the fast response.  The original command give back an empty mrc array, but once I remove --center flag it runs without errors and returns rescaled template as expected. What are the consistences of omitting `--center` flag for the template matching?  
Best, 
Oleksiy

**Comment by McHaillet** [comment]
Hi!
Maybe this centering option should be reevaluated. I had some small issues with it before. This is the scipy docs for center of mass (https://docs.scipy.org/doc/scipy/reference/generated/scipy.ndimage.center_of_mass.html) the bottom lines say it can return a RuntimeWarning and return inf or nan values, which might be why your template ends up empty.

> What are the consistences of omitting `--center` flag for the template matching?

It wont matter a lot, unless your structure is really in a corner. In that case it might rotate out of view as the center of rotation for the template is always chosen as the center of the box. But if your template is already approximately centered it won't matter.

You can alwasy use chimera to recenter manually. If you open the mask and the prepared template. You can drag the template into the center of the mask and then resample it to the grid of the mask (something like `vop resample #1 onGrid #2`). Again though, if already approx. centered, I would not bother!

Best,
Marten

**Reply by kovtunlab** [reply]
Thank you a lot for the explainer. I will recenter the particle manually then in chimera. Good to clarify the centering is jsut a relative position in the box, not some parameters in the mrc header. 

Best, 

Oleksiy

**Comment by frozenfas** [comment]
I also encountered this error when using the centre option. It disappeared when I updated Pytom.  I mean, I had a problem in the version of pytom_create_template.py, where you could give CTF parameters (Cs, kv, etc), but in the recent version, the problem solved itself.

**Reply by McHaillet** [reply]
Good to hear that it works after the update! (see my comment below)

**Comment by McHaillet** [comment]
With the 0.7.0 release the centering issue has been fixed, see PR #179.
